common/persistence/cassandra/queue_store_v2.go

```diff
@@ -0,0 +1,356 @@
+	}
+
+	return messageID, nil
+}
+
+func (q *QueueStoreV2) getLastMessageID(
+	ctx context.Context,
+	request persistence.InternalEnqueueMessageRequest,
+	partition int,
+) (int64, error) {
+	query := q.session.Query(templateGetLastMessageIDQuery, request.QueueType, request.QueueName, partition).WithContext(ctx)
+	result := make(map[string]interface{})
+	err := query.MapScan(result)
+	if err != nil {
+		if gocql.IsNotFoundError(err) {
+			return persistence.EmptyQueueMessageID, nil
+		}
+		return persistence.EmptyQueueMessageID, gocql.ConvertError("getLastMessageID", err)
+	}
+	return result["message_id"].(int64), nil
+}
+
+func (q *QueueStoreV2) getCurrentAckLevelAndVersion(ctx context.Context, queueType persistence.QueueV2Type, queueName string) (ackLevel int64, version int64, err error) {
+	query := q.session.Query(templateGetCurrentAckLevelAndVersionQueryV2, queueType, queueName).WithContext(ctx)
+	result := make(map[string]interface{})
+	err = query.MapScan(result)
+	if err != nil {
+		if gocql.IsNotFoundError(err) {
+			return 0, 0, nil
+		}
+		return 0, 0, gocql.ConvertError("getLastMessageID", err)
+	}
+	version = result["version"].(int64)
+	payload := result["metadata_payload"].([]byte)
+	enconding := result["metadata_encoding"].(string)
+
+	ackLevel, err = getAckLevelFromQueueV2Metadata(payload, enconding)
+	if err != nil {
+		return 0, 0, fmt.Errorf("failed to get ack level from queue metadata: %w, Type: %v, Name: %v", err, queueType, queueName)
+	}
+	return ackLevel, version, nil
+}
+
+func (q *QueueStoreV2) updateAckLevel(ctx context.Context, request persistence.InternalRangeDeleteMessagesRequest, version int64) error {
+	queueMetadata := QueueV2MetadataPayload{
+		AckLevel: request.InclusiveMaxMessageMetadata.ID,
+	}
+	blob, err := convertQueueV2MetadataToBlob(&queueMetadata)
+	if err != nil {
+		return fmt.Errorf("failed to update ack level: %w, Type: %v, Name: %v", err, request.QueueType, request.QueueName)
+	}
+
+	for {
+		query := q.session.Query(templateUpdateQueueMetadataQueryV2,
+			blob.Data,
+			blob.EncodingType.String(),
+			version+1, // always increase version number on update
+			request.QueueType,
+			request.QueueName,
+			version, // condition update
+		).WithContext(ctx)
+
+		result := make(map[string]interface{})
+		applied, err := query.MapScanCAS(result)
+		if err != nil {
+			return gocql.ConvertError("updateAckLevel", err)
+		}
+		if applied {
+			break
+		}
+		version = result["version"].(int64)
+		payload := result["metadata_payload"].([]byte)
+		enconding := result["metadata_encoding"].(string)
+
+		ackLevel, err := getAckLevelFromQueueV2Metadata(payload, enconding)
+		if err != nil {
+			return fmt.Errorf("failed to get ack level from queue metadata: %w, Type: %v, Name: %v", err, request.QueueType, request.QueueName)
+		}
+		if ackLevel >= request.InclusiveMaxMessageMetadata.ID {
+			break
+		}
+		// otherwise repeat until we get it right
+		// TODO: do we need to limit the number of retries?
```

Comment:
yeah I think we should, and it's ok for the ack level to be not up to date.
With current impl, looks like it will loop until context timeout.